rfcs/2021-09-01-8547-accept-metrics-in-datadog-agent-source.md

```diff
@@ -0,0 +1,192 @@
+    - Add support for missing fields in the `datadog_metrics` sinks
+    - Align tag manipulation (Datadog allows `key:foo` & `key:bar`, and the tags are encoded as a list whereas Vector
+      encode those as a map).
+    - Overall this is fairly straighforward except for the tagging issue
+- Handle the `/api/beta/sketches` route in the `datadog_agent` source to support sketches/distribution encoded using
+  protobuf
+   - Revamp the distribution metrics in the `datadog_metrics` sink to use sketches and the associated endpoint will
+     probably be a good idea.
+   - The sketches the agent ships is based on this [paper](http://www.vldb.org/pvldb/vol12/p2195-masson.pdf) whereas
+     Vector uses what's called a summary inside the Agent, implementing the complete DDSketch support in Vector is
+     probably a good idea as sketches have convenient properties for wide consistent aggregation and limited error. To
+     support smooth migration, full DDsktech support is mandatory, as customers that emit distribution metric from
+     Datadog Agent would need it to migrate to Vector aggegation.
+
+**Regarding the tagging issue:** A -possibly temporary- work-around would be to store incoming tags with the complete
+"key:value" string as the key and an empty value to store those in the extisting map Vector uses to store
+[tags](https://github.com/timberio/vector/blob/master/lib/vector-core/src/event/metric.rs#L60) and slightly rework the
+`datadog_metrics` sink not to append `:` if a tag key has the empty string as the corresponding value.
+
+
+## Rationale
+
+- Smoother Vector integration with Datadog.
+- Needed for Vector to act as a complete Datadog Agent aggregator (but further work will still be required).
+- Extend the Vector ecosystem, bring additional feature for distribution metrics that would enable consistent
+  aggregation.
+
+## Drawbacks
+
+- It arbitrary imposes internal sketches representation to follow the DDSketch paper.
+
+## Prior Art
+
+- Existing Vector metrics source over HTTP(s) works well, those constructs will be reused
+- Datadog logs is already supported in the `datadog_agent` source
+- Vector has its own vector-to-vector protocol that serve a similar purpose but the idea is to enable Vector to receive
+  Datadog metrics from unmodified Datadog Agents so that is probably not an option
+- There are many [ways](https://edoliberty.github.io/papers/streamingQuantiles.pdf) of implementing quantile sketches
+- Opentelemetry is moving to official support sketches on a [general
+  basis](https://github.com/open-telemetry/opentelemetry-specification/issues/982).
+
+## Alternatives
+
+- Use an alternate protocol between Datadog Agents and Vector (Like Prometheus, Statds or Vector own protocol)
+- For sketches, we could flatten sketches and compute usual derived metrics (min/max/average/count/some percentile) and
+  send those as gauge/count, but it would prevent (or at least impact) existing distribution/sketches user
```

Comment:
Indeed, and this would hugely impact UX because sketches are tagged and this allows users to filter a distribution metrics in the DD app and show up an accurate distribution only for the matching tags, without sketches this would not be possible. I will add that to the RFC.